python/mspasspy/db/database.py

```diff
@@ -187,91 +192,153 @@ def read_data(self, object_id, load_history=False, include_undefined=False, excl
+
+        elif mode == "pedantic":
+            for k in md:
+                if read_metadata_schema.is_defined(k):
+                    if not isinstance(md[k], read_metadata_schema.type(k)):
+                        fatal_keys.append(k)
+                        is_dead = True
+                        log_error_msg.append("pedantic mode: {} has type {}, forbidden by definition".format(k, type(md[k])))
+
+
+        # 1.4 create a mspass object by passing MetaData
+        # if not changing the fatal key values, runtime error in construct a mspass object
+        for k in fatal_keys:
+            if read_metadata_schema.type(k) is str:
+                md[k] = ""
+            elif read_metadata_schema.type(k) is int:
+                md[k] = 0
+            elif read_metadata_schema.type(k) is float:
+                md[k] = 0.0
+            elif read_metadata_schema.type(k) is bool:
+                md[k] = False
+            elif read_metadata_schema.type(k) is dict:
+                md[k] = {}
+            elif read_metadata_schema.type(k) is list:
+                md[k] = []
+            elif read_metadata_schema.type(k) is bytes:
+                md[k] = b'\x00'
+            else:
+                md[k] = None
 
         if object_type is TimeSeries:
             # FIXME: This is awkward. Need to revisit when we have proper constructors.
             mspass_object = TimeSeries({k: md[k] for k in md}, np.ndarray([0], dtype=np.float64))
             # FIXME: if npts is in the exclude list or not in the schema, the following won't work.
             # May need to consider adding a "required" key to the metadata schema to avoid invalid combination.
-            mspass_object.npts = object_doc['npts']
+            if 'npts' in object_doc:
+                mspass_object.npts = object_doc['npts']
         else:
             mspass_object = Seismogram(_CoreSeismogram(md, False))
 
-        # 2.load data from different modes
-        mode = object_doc['storage_mode']
-        if mode == "file":
-            self._read_data_from_dfile(mspass_object, object_doc['dir'], object_doc['dfile'], object_doc['foff'])
-        elif mode == "gridfs":
-            self._read_data_from_gridfs(mspass_object, object_doc['gridfs_id'])
-        elif mode == "url":
-            pass  # todo for future
-        else:
-            raise TypeError("Unknown storage mode: {}".format(mode))
-
-        # 3.load history
-        if load_history:
-            history_obj_id_name = self.database_schema.default_name('history_object') + '_id'
-            if history_obj_id_name in object_doc:
-                self._load_history(mspass_object, object_doc[history_obj_id_name])
+        # not continue step 2 & 3 if the mspass object is dead
+        if is_dead:
+            mspass_object.kill()
+            for msg in log_error_msg:
+                mspass_object.elog.log_error('read_data', msg, ErrorSeverity.Invalid)
+            self._save_elog(mspass_object)
```

Comment:
We shouldn't save the object to elog's graveyard within a read method. That mechanism is designed for save only. Remove this line should fix it.